[BEAM-5191] Support for BigQuery clustering

jklukas opened the pull request:
This takes the commits from #7061, rebases on master, and adds an `enableClustering` method to allow users to opt in to the updated coder when using dynamic destinations.

This follows the pattern of #6914 where we added a new version of `MetadataCoder`, documenting how a user could opt in and why they might want to.

This should solve the coder compatibility concerns that were the blockers to merging #7061. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

jklukas commented:
R: @juancho088 @reuvenlax @alexvanboxel @chamikaramj who were reviewers on #7061

Also cc @wscheep who authored #7061 which is still the bulk of the code here.

jklukas commented:
Run JavaPortabilityApi PreCommit

jklukas commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:
```diff
@@ -391,6 +392,93 @@ public void testTimePartitioning(BigQueryIO.Write.Method insertMethod) throws Ex
+                .withTestServices(fakeBqServices)
+                .withMethod(BigQueryIO.Write.Method.FILE_LOADS)
+                .withSchema(schema)
+                .enableClustering()
```
Notably, this test fails if `enableClustering()` is not called because the default `TableDestinationCoderV2` is used and clustering information is dropped before the table is created. This is exactly the behavior we want for backwards compatibility.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTimePartitioningClusteringIT.java:
```diff
@@ -0,0 +1,150 @@
+    p.apply(BigQueryIO.readTableRows().from(options.getInput()))
+        .apply(ParDo.of(new KeepStationNumberAndConvertDate()))
+        .apply(
+            BigQueryIO.writeTableRows()
```
Can we also add one with dynamic destinations and clustering ?

jklukas replied:
Added. It doesn't look like the ITs are run by Jenkins by default and I also can't find that there's a command to kick off BigQueryIO integration tests. Do you have any advice about how we can get results for these?

chamikaramj replied:
Post-commit test suite should capture these. Just triggered it.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:
```diff
@@ -1793,6 +1814,32 @@ static String getExtractDestinationUri(String extractDestinationDir) {
+     * option, since {@link TableDestinationCoderV3} will not be able to read state written with a
+     * previous version.
+     */
+    public Write<T> enableClustering() {
```
Having these two methods seems to make the API pretty brittle.

How about just having one function withClustering() that optionally takes a clustering object ? In dynamic destinations case optional Clustering object can be skipped/null and the method will behave similar to enableClustering() ?

jklukas replied:
Implemented. `enableClustering()` is now `withClustering()` and it sets `clustering` to a default `Clustering` instance with no fields set, which we use as a flag for enabling clustering on dynamic destinations.

jklukas commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinationsHelpers.java:
```diff
@@ -167,7 +175,12 @@ public TableDestination getTable(DestinationT destination) {
     Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
         throws CannotProvideCoderException {
-      return inner.getDestinationCoderWithDefault(registry);
+      Coder<DestinationT> destinationCoder = getDestinationCoder();
```
`DynamicDestinations#getDestinationCoderWithDefault` is commented as:

```
  // Gets the destination coder. If the user does not provide one, try to find one in the coder
  // registry. If no coder can be found, throws CannotProvideCoderException.
```

This code is written with potentially multiple layers of delegation, and I think the correct behavior here is to return the first non-delegated implementation of `getDestinationCoder()` that appears as we move down the delegation chain.

I would argue that the existing behavior is incorrect. Currently, if an implementing class defines a custom return value for `getDestinationCoder`, that value is ignored when you call `getDestinationCoderWithDefault`. My expectation is that `getDestinationCoderWithDefault` would always return the same value as `getDestinationCoder` except in the null case, in which `getDestinationCoderWithDefault` would then attempt to look up a coder in the registry.

So the change here is intended to fix broken behavior. 

It's possible that a user has written a custom class that extends DelegatingDynamicDestinations and relies on the incorrect behavior, but it feels unlikely to me.

For the scope of the coders provided here, I don't believe this change affects behavior (the method was already returning TableDestinationCoderV2 in all cases).

jklukas replied:
It may actually be better to remove this override altogether. `getDestinationCoderWithDefault` is defined as package-private and probably shouldn't be overriden. Instead, we let `getDestinationCoder` handle delegation, and the base implementation of `getDestinationCoderWithDefault` will make the appropriate `getDestinationCoder` call before falling back to the registry.

jklukas replied:
I've removed the override getDestinationCoderWithDefault completely.

jklukas replied:
> I've removed the override getDestinationCoderWithDefault completely.

Looks like that caused some regressions where destination coder couldn't be found from type. I'm looking into it.

jklukas replied:
The override turns out to be necessary because the `DynamicDestinations` instance itself is eventually passed into `extractFromTypeParameters` and the inner class potentially has richer type information than the delegating class.

jklukas replied:
I've put the override method back in place, including the modification of first checking if we have a non-null return from `getDestinationCoder()`.

jklukas commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTimePartitioningClusteringIT.java:
```diff
@@ -0,0 +1,150 @@
+                .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_IF_NEEDED)
+                .withWriteDisposition(BigQueryIO.Write.WriteDisposition.WRITE_TRUNCATE));
+
+    p.run().waitUntilFinish();
```
I believe these ITs are the work of @wscheep in the previous PR, and I am going to borrow liberally from these examples to add the dynamic destinations ITs.

chamikaramj commented:
Run Java PostCommit

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinationsHelpers.java:
```diff
@@ -165,8 +173,12 @@ public TableDestination getTable(DestinationT destination) {
 
     @Override
-    Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
+    public Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
```
Was change of modifier here (to public) intentional ? Wasn't sure from your explanation below.

jklukas replied:
Unintentional. Will revert.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:
```diff
@@ -391,6 +392,93 @@ public void testTimePartitioning(BigQueryIO.Write.Method insertMethod) throws Ex
   }
 
+  @Test
+  public void testClusteringStreamingInserts() throws Exception {
```
Probably also add a test that confirms that we fail if withClustering() is set without withTimePartitioning().

jklukas replied:
Added

```
  @Test(expected = IllegalArgumentException.class)
  public void testClusteringThrowsWithoutPartitioning() throws Exception {
```

chamikaramj commented:
FYI: I'm OOO till 6/9 so please add another reviewer if you need to get this merged early.

LGTM from my side other than the two comments I mentioned and integration tests passing.

jklukas commented:
Run Java PostCommit

jklukas commented:
> LGTM from my side other than the two comments I mentioned and integration tests passing.

Addressed those comments, integration tests passed, and now I've squashed and rebased on master.

> FYI: I'm OOO till 7/9 so please add another reviewer if you need to get this merged early.

I'm also OOO next week, so if there are any further changes that need to be made, I'll be back on 7/15. We have several weeks before the branch cut date for 2.15, so I don't see any point in rushing. 

And thanks for the thoughtful review, @chamikaramj .

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Java PostCommit

reuvenlax commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoderV3.java:
```diff
@@ -0,0 +1,64 @@
+ * A {@link Coder} for {@link TableDestination} that includes time partitioning and clustering
+ * information. Users must opt in to this version of the coder by setting one of the clustering
+ * options on {@link BigQueryIO.Write}, otherwise {@link TableDestinationCoderV2} will be used and
+ * clustering information will be discarded.
```
if a user forgets to set a clustering option but creates a TableDestination with clustering information, will we warn them or just silently discard?

jklukas commented:
There is no warning at this point. For the case of loading to an existing
table, the operation will throw exceptions anyway if clustering is
mismatched.  Perhaps there are compile time cases we could warn about.

I will have to think more about when I'm back in office next week if we
want to block on that.

On Mon, Jul 8, 2019, 1:06 PM reuvenlax <notifications@github.com> wrote:

> *@reuvenlax* commented on this pull request.
> ------------------------------
>
> In
> sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoderV3.java
> <https://github.com/apache/beam/pull/8945#discussion_r301203857>:
>
> > + */
> +package org.apache.beam.sdk.io.gcp.bigquery;
> +
> +import java.io.IOException;
> +import java.io.InputStream;
> +import java.io.OutputStream;
> +import org.apache.beam.sdk.coders.AtomicCoder;
> +import org.apache.beam.sdk.coders.Coder;
> +import org.apache.beam.sdk.coders.NullableCoder;
> +import org.apache.beam.sdk.coders.StringUtf8Coder;
> +
> +/**
> + * A {@link Coder} for {@link TableDestination} that includes time partitioning and clustering
> + * information. Users must opt in to this version of the coder by setting one of the clustering
> + * options on {@link BigQueryIO.Write}, otherwise {@link TableDestinationCoderV2} will be used and
> + * clustering information will be discarded.
>
> if a user forgets to set a clustering option but creates a
> TableDestination with clustering information, will we warn them or just
> silently discard?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/8945?email_source=notifications&email_token=AAFIW5DW44XOYCDV56PR263P6NXZFA5CNFSM4H3LPCH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5YJ5IY#pullrequestreview-259038883>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFIW5HCBJWDJJP3EH4GJE3P6NXZFANCNFSM4H3LPCHQ>
> .
>

reuvenlax commented:
A "good enough" solution might be to always print a warning when clustering
is not enabled. We will want to wordsmith the warning so it's not too
frightening, and it's clear that it only applies if trying to specify
clustering.

On Mon, Jul 8, 2019 at 1:11 PM Jeff Klukas <notifications@github.com> wrote:

> There is no warning at this point. For the case of loading to an existing
> table, the operation will throw exceptions anyway if clustering is
> mismatched. Perhaps there are compile time cases we could warn about.
>
> I will have to think more about when I'm back in office next week if we
> want to block on that.
>
> On Mon, Jul 8, 2019, 1:06 PM reuvenlax <notifications@github.com> wrote:
>
> > *@reuvenlax* commented on this pull request.
> > ------------------------------
> >
> > In
> >
> sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoderV3.java
> > <https://github.com/apache/beam/pull/8945#discussion_r301203857>:
> >
> > > + */
> > +package org.apache.beam.sdk.io.gcp.bigquery;
> > +
> > +import java.io.IOException;
> > +import java.io.InputStream;
> > +import java.io.OutputStream;
> > +import org.apache.beam.sdk.coders.AtomicCoder;
> > +import org.apache.beam.sdk.coders.Coder;
> > +import org.apache.beam.sdk.coders.NullableCoder;
> > +import org.apache.beam.sdk.coders.StringUtf8Coder;
> > +
> > +/**
> > + * A {@link Coder} for {@link TableDestination} that includes time
> partitioning and clustering
> > + * information. Users must opt in to this version of the coder by
> setting one of the clustering
> > + * options on {@link BigQueryIO.Write}, otherwise {@link
> TableDestinationCoderV2} will be used and
> > + * clustering information will be discarded.
> >
> > if a user forgets to set a clustering option but creates a
> > TableDestination with clustering information, will we warn them or just
> > silently discard?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/apache/beam/pull/8945?email_source=notifications&email_token=AAFIW5DW44XOYCDV56PR263P6NXZFA5CNFSM4H3LPCH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5YJ5IY#pullrequestreview-259038883
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAFIW5HCBJWDJJP3EH4GJE3P6NXZFANCNFSM4H3LPCHQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/8945?email_source=notifications&email_token=AFAYJVMDVF3DPJ4JDFBEIJLP6ONN3A5CNFSM4H3LPCH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZOGXUY#issuecomment-509373395>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFAYJVKO7SCMPFTQGZWTTLTP6ONN3ANCNFSM4H3LPCHQ>
> .
>

chamikaramj commented:
Thanks. I agree that a compile time (or even run-time) failure is better but a warning should be OK if this is not possible for some reason.

I'll wait till this is addressed for the merge.

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Java PreCommit

jklukas commented:
Run JavaPortabilityApi PreCommit

jklukas commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTables.java:
```diff
@@ -130,6 +130,16 @@ private TableDestination getTableDestination(ProcessContext context, Destination
+              + " to use TableDestinationCoderV2. Set withClustering() on BigQueryIO.write() and, "
+              + " if you provided a custom DynamicDestinations instance, override"
+              + " getDestinationCoder() to return TableDestinationCoderV3.",
+          dynamicDestinations);
```
We cannot know at compile-time whether a custom table function or DynamicDestinations instance will produce any destination with clustering enabled, so we have to check the produced destinations at runtime.

This check will cause a runtime failure for StreamingInserts if a clustered destination is produced without the relevant configuration to use the newer destination coder.

jklukas commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:
```diff
@@ -165,6 +166,16 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
+              + " to use TableDestinationCoderV2. Set withClustering() on BigQueryIO.write() and, "
+              + " if you provided a custom DynamicDestinations instance, override"
+              + " getDestinationCoder() to return TableDestinationCoderV3.",
+          dynamicDestinations);
```
This check will cause a runtime failure for BatchLoads if a clustered destination is produced without the relevant configuration to use the newer destination coder.

jklukas commented:
R: @chamikaramj 

I've added in runtime checks for both the streaming insert and batch load cases where the transform will fail with an exception that includes instructions about how to configure clustering if a clustered destination is produced without the right coder in place.

Does this look ready to merge?

chamikaramj commented:
LGTM. Thanks Jeff.

chamikaramj commented:
Run Java PostCommit

chamikaramj commented:
Run Dataflow ValidatesRunner

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Java PostCommit

jklukas commented:
Run Dataflow ValidatesRunner

jklukas commented:
@chamikaramj All the tests here are finally passing again after transient issues.

chamikaramj commented:
Thanks. Merging.
